deploy/libproject/deploy.rb

```diff
@@ -0,0 +1,207 @@
+require "open3"
+require "set"
+require_relative "../../aou-utils/serviceaccounts"
+require_relative "../../aou-utils/utils/common"
+require_relative "../../aou-utils/workbench"
+require_relative "../../api/libproject/devstart"
+require_relative "../../api/libproject/gcloudcontext"
+require_relative "../../api/libproject/wboptionsparser"
+
+
+DOCKER_KEY_FILE_PATH = "/creds/sa-key.json"
+
+def get_source_project(p)
+  # Returns the project preceding "p" in the release promotion process.
+  return {
+    "all-of-us-rw-stable" => "all-of-us-rw-staging"
+  }[p]
+end
+
+def capture_stdout2(cmd)
+  out, _ = Open3.capture2(*cmd)
+  return out
+end
+
+def get_live_gae_version(project)
+  versions = capture_stdout2 %W{
+    gcloud app
+    --format json(id,service,traffic_split)
+    --project #{project}
+    versions list
+  }
+  if versions.empty?
+    Common.new.error "Failed to get live version from gcloud"
+    exit 1
+  end
+  services = Set["api", "default", "public-api"]
+  actives = JSON.parse(versions).select{|v| v["traffic_split"] == 1.0}
+  if actives.empty?
+    return nil
+  elsif actives.length > services.length
+    # Found some extra services. Log
+    return nil
+  elsif services != actives.map{|v| v["service"]}.to_set
+    # Found the wrong services
+    return nil
+  end
+
+  versions = actives.map{|v| v["id"]}.to_set
+  if versions.length != 1
```

Comment:
So basically if we somehow have different versions for two of api, default, and public-api we will bail out here and require the client to specify a version?

You may want to add a comment, or a log message when this happens.

Reply:
Yep. Done.